src/worker/network.rs

```diff
@@ -200,7 +202,22 @@ impl NetworkWorker {
                 let lifetime: f64 =
                     thread_rng().sample(Exp::new(departure_rate).unwrap());
 
-                dynamic_sockets.insert(handle, (SystemTime::now(), lifetime));
+                // If we've reached the connections limit
+                if dynamic_sockets.len() == n_dyn_connections as usize
+                    && preempt
+                {
+                    let idx =
+                        thread_rng().gen_range(0..n_dyn_connections as usize);
+                    let (key, _) = sockets.iter().nth(idx).unwrap();
+                    dynamic_sockets.remove(&key);
+                    close_sockets.push(key);
+
+                    dynamic_sockets
+                        .insert(handle, (SystemTime::now(), lifetime));
+                } else {
+                    dynamic_sockets
+                        .insert(handle, (SystemTime::now(), lifetime));
+                }
```

Comment:
I don't think the premise of `preempt` holds here, If it's false a new socket is always added, think we might want something like this:
```suggestion
                if dynamic_sockets.len() == n_dyn_connections as usize
                {
                    if preempt {
                        let idx =
                            thread_rng().gen_range(0..n_dyn_connections as usize);
                        let (key, _) = sockets.iter().nth(idx).unwrap();
                        dynamic_sockets.remove(&key);
                        close_sockets.push(key);

                        dynamic_sockets
                            .insert(handle, (SystemTime::now(), lifetime));
                    }
                } else {
                    dynamic_sockets
                        .insert(handle, (SystemTime::now(), lifetime));
                }
```

Reply:
ahh yeah good point, my bad